plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx

```diff
@@ -109,3 +119,77 @@ export const addProductCollectionBlockToParentOfPaginationBlock = () => {
 		);
 	}
 };
+
+export const useSetPreviewState = ( {
+	setPreviewState,
+	location,
+	attributes,
+	setAttributes,
+}: {
+	setPreviewState?: SetPreviewState | undefined;
+	location: WooCommerceBlockLocation;
+	attributes: ProductCollectionAttributes;
+	setAttributes: (
+		attributes: Partial< ProductCollectionAttributes >
+	) => void;
+} ) => {
+	const setState = ( newPreviewState: PreviewState ) => {
+		setAttributes( {
+			__privatePreviewState: {
+				...attributes.__privatePreviewState,
+				...newPreviewState,
+			},
+		} );
+	};
+
+	// Running setPreviewState function provided by Collection, if it exists.
+	useLayoutEffect( () => {
+		if ( ! setPreviewState ) {
+			return;
+		}
+
+		const cleanup = setPreviewState?.( {
+			setState,
+			location,
+			attributes,
+		} );
+
+		if ( cleanup ) {
+			return cleanup;
+		}
+
+		// It should re-run only when setPreviewState changes to avoid performance issues.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [ setPreviewState ] );
+
+	/**
+	 * For all Product Collection blocks that inherit query from the template,
+	 * we want to show a preview message in the editor if the block is in
+	 * generic archive template i.e.
+	 * - Products by category
+	 * - Products by tag
+	 * - Products by attribute
+	 */
+	useLayoutEffect( () => {
```

Comment:
I assume this piece will be deprecated once we establish a proper way of handling preview using a new argument like `uses_context` in the `registerProductCollection` function. Is that right?

Reply:
We may or we may not. Let's assume; we will deprecate this. In that case, we will need to move the logic to calculate `isGenericArchiveTemplate` and then set `uses_context` if isGenericArchiveTemplate is `true`. Therefore, if we have to use the same logic somewhere else, then maybe we can keep it here instead. But yes, we can definitely replace it with `uses_context` if required. Maybe we can discuss this in more detail while implementing `uses_context` . What do you think? 🙂

Reply:
I think it's best to differentiate the `uses_context` feature, which is a core function, from the 3PD feature in this API. This is important because it allows 3PDs to function as expected when a specific context is requested in the `uses_context` arg. 
But we can discuss this further as needed. I just wanted to flag that this code part will need some refactoring work right after. 

Consider the following scenario: 
- A 3PD creates a collection and sets up a preview handler to check the availability of API-based recommendations **for a specific product**.
- The `uses_context` argument is specified to ensure that each preview request coming from the Product Collection block includes the productID. 
- The developer expects that every time the preview handler runs, **it always has the specific product ID** (in order to check.) 

At least, this is how the DX works in my mind, I love to hear your thoughts as well @imanish003 cc @kmanijak

Reply:
Aha! This is a new insight for me i.e. we want to run preview handler only when, whatever is provided inside `uses_context` is available. In your scenario, when product ID is available. I agree & it makes sense 💯 

But still, in case of:
1. Products by category
2. Products by tag
3. Products by attribute

We know that the collection will always be in preview mode if *Inherit query from template is true*. As you can see [this](https://github.com/woocommerce/woocommerce/blob/6c6181f340d51aa76b328570567816ad3f5d9095/plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx#L175) if condition that we run logic only if `setPreviewState` isn't provided. So I believe [this logic](https://github.com/woocommerce/woocommerce/blob/6c6181f340d51aa76b328570567816ad3f5d9095/plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx#L174-L197) can still remain same when we will implement `uses_context`. But we can think about it more while implementing `uses_context`. While testing it, we may come across more scenarios. Let's discuss it in more details while implementing `uses_context`. If this sounds good to you, we can resolve this conversation 🙂

Reply:
Sounds good! I just wanted to bring up some ideas for consideration in the next iteration.